README.md
# Thunderzippy

Secure streaming zipped download bundles.

A microservice that takes a list of remote files and provides a url that streams
a zipped bundle of all those files.

Concept based on an idea from
http://engineroom.teamwork.com/how-to-securely-provide-a-zip-download-of-a-s3-file-bundle/
then simplified to work as a simple RESTful API.

## Install and Run

```
docker-compose build
docker-compose up
```

### Environment Variables

`REDIS_ADDRESS` : Address with port of Redis host e.g. `127.0.0.1:6379`
`PORT` : Local port the HTTP server will bind to

## Use

### Create zip
```
$ curl -H "Content-Type: application/json" -X POST -d '[{"Filepath":"images/example.jpg","Url":"https://upload.wikimedia.org/wikipedia/mediawiki/a/a9/Example.jpg"}]' http://localhost:8080/zip/

{"ref":"z95h07"}
```

### Download zip
```
$ curl http://localhost:8080/zip/?ref=z95h07 -o "download.zip"
$ unzip -l download.zip
Archive:  download.zip
  Length     Date   Time    Name
 --------    ----   ----    ----
    61136  02-26-16 03:00   images/example.jpg
 --------                   -------
    61136                   1 file
```

### License

Thunderzippy is released under the [MIT License](http://www.opensource.org/licenses/MIT).

Patches, suggestions and comments are welcome.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).


## [v2.0] - 2018-06-24
### Added
- Docker and docker-compose files
### Changed
- Replaced config file with env vars
- Updated redigo upstream

## [v1.0] - 2016-06-03
### Added
- Initial release


db.go
package main

import (
  "encoding/json"
  "strconv"
  "log"
  "time"
  "errors"
  "math/rand"
  "os"
  redigo "github.com/gomodule/redigo/redis"
)

var redisPool *redigo.Pool

func init() {
  log.Printf("Opening Redis: %s", os.Getenv("REDIS_ADDRESS"))
  redisPool = &redigo.Pool{
    MaxIdle:     10,
    IdleTimeout: 1 * time.Second,
    Dial: func() (redigo.Conn, error) {
      return redigo.Dial("tcp", os.Getenv("REDIS_ADDRESS"))
    },
    TestOnBorrow: func(c redigo.Conn, t time.Time) (err error) {
      _, err = c.Do("PING")
      if err != nil {
          panic("Error connecting to redis")
      }
      return
    },
  }
}

func getFileListByZipReferenceId(id string) (files []*ZipEntry, err error) {
  redis := redisPool.Get()
  defer redis.Close()

  // Get the value from Redis
  result, err := redis.Do("GET", "zip:" + id)
  if err != nil || result == nil {
    err = errors.New("Access Denied (sorry your link has timed out)")
    return
  }

  // Convert to bytes
  var resultByte []byte
  var ok bool
  if resultByte, ok = result.([]byte); !ok {
    err = errors.New("Error converting data stream to bytes")
    return
  }

  // Decode JSON
  if err = json.Unmarshal(resultByte, &files); err != nil {
    err = errors.New("Error decoding json: " + string(resultByte))
  }
  return
}

func CreateZipReference(files []*ZipEntry) (ref_id_string string) {
  redis := redisPool.Get()
  defer redis.Close()

  filesJson, err := json.Marshal(files)
  HandleError(err)

  //get new id redis
  ref_id, err := redis.Do("INCR", "zip_reference_id")
  HandleError(err)
  ref_id_string = RandomString(5) + strconv.FormatInt(ref_id.(int64), 10)

  // Save JSON files to Redis
  _, err = redis.Do("SET", "zip:" + ref_id_string, filesJson)
  HandleError(err)

  return
}

func RandomString(strlen int) string {
  rand.Seed(time.Now().UTC().UnixNano())
  const chars = "abcdefghijklmnopqrstuvwxyz0123456789"
  result := make([]byte, strlen)
  for i := 0; i < strlen; i++ {
    result[i] = chars[rand.Intn(len(chars))]
  }
  return string(result)
}


zippy.go
package main

import (
    "archive/zip"
    "encoding/json"
    "errors"
    redigo "github.com/garyburd/redigo/redis"
    "io"
    "log"
    "net/http"
    "os"
    "time"
)

var redisPool *redigo.Pool
var config = Configuration{}

type Configuration struct {
    RedisServerAndPort string
    Port               string
}

type ZipEntry struct {
    Filepath, Url string
}

func zipHandler(w http.ResponseWriter, r *http.Request) {
    log.Printf("%s\t\t%s", r.Method, r.RequestURI)
    start := time.Now()

    refs, ok := r.URL.Query()["ref"]
    if !ok || len(refs) < 1 {
        http.Error(w, "Thunderzippy. Pass ?ref= to use.", 500)
        return
    }
    ref := refs[0]

    files, err := getFileListFromRedis(ref)
    if err != nil {
        http.Error(w, err.Error(), 403)
        log.Printf("%s\t%s\t%s", r.Method, r.RequestURI, err.Error())
        return
    }

    w.Header().Add("Content-Disposition", "attachment; filename=\"test.zip\"")
    w.Header().Add("Content-Type", "application/zip")

    zipWriter := zip.NewWriter(w)

    for _, file := range files {
        addDownloadToZip(zipWriter, file.Url, file.Filepath)
    }

    err = zipWriter.Close()
    if err != nil {
        log.Fatal(err)
    }

    log.Printf("Thunderzipped:\t%d files (%s)", len(files), time.Since(start))
}

func addDownloadToZip(zipWriter *zip.Writer, url string, name string) {
    // https://golang.org/pkg/net/http/
    resp, err := http.Get(url)
    if err != nil {
        log.Print(err)
        return
    }
    log.Printf("adding:\t%d %s", resp.StatusCode, url)
    if resp.StatusCode != 200 {
        return
    }
    defer resp.Body.Close()

    // https://golang.org/pkg/archive/zip/#FileHeader
    h := &zip.FileHeader{
        Name:   name,
        Method: zip.Deflate,
        Flags:  0x800,
    }
    h.SetModTime(time.Now())
    f, err := zipWriter.CreateHeader(h)
    if err != nil {
        log.Print(err)
        return
    }

    io.Copy(f, resp.Body)
}

func getFileListFromRedis(ref string) (files []*ZipEntry, err error) {
    redis := redisPool.Get()
    defer redis.Close()

    // Get the value from Redis
    result, err := redis.Do("GET", "zip:"+ref)
    if err != nil || result == nil {
        err = errors.New("Access Denied (sorry your link has timed out)")
        return
    }

    // Convert to bytes
    var resultByte []byte
    var ok bool
    if resultByte, ok = result.([]byte); !ok {
        err = errors.New("Error converting data stream to bytes")
        return
    }

    // Decode JSON
    err = json.Unmarshal(resultByte, &files)
    if err != nil {
        err = errors.New("Error decoding json: " + string(resultByte))
    }

    return
}

func initRedis() {
    log.Printf("Opening Redis: %s", config.RedisServerAndPort)
    redisPool = &redigo.Pool{
        MaxIdle:     10,
        IdleTimeout: 1 * time.Second,
        Dial: func() (redigo.Conn, error) {
            return redigo.Dial("tcp", config.RedisServerAndPort)
        },
        TestOnBorrow: func(c redigo.Conn, t time.Time) (err error) {
            _, err = c.Do("PING")
            if err != nil {
                panic("Error connecting to redis")
            }
            return
        },
    }
}

func initConfig() {
    conf := "thunderzippy_conf.json"
    configFile, _ := os.Open(conf)
    log.Printf("Reading config %s", conf)
    decoder := json.NewDecoder(configFile)
    err := decoder.Decode(&config)
    if err != nil {
        panic("Error reading conf")
    }
}

func main() {
    initConfig()
    initRedis()
    log.Printf("Thunderzippy is go on port %s", config.Port)
    http.HandleFunc("/zip/", zipHandler)
    http.ListenAndServe(":"+config.Port, nil)
}


handler.go
package main

import (
  "archive/zip"
  "encoding/json"
  "net/http"
  "io"
  "io/ioutil"
  "time"
  "log"
)

func ZipReferenceHandler(w http.ResponseWriter, r *http.Request) {
  switch r.Method {
    case "GET":
      ZipReferenceDownload(w, r)
    case "POST":
      ZipReferenceCreate(w, r)
  }
}

func ZipReferenceDownload(w http.ResponseWriter, r *http.Request) {
    log.Printf("%s\t\t%s", r.Method, r.RequestURI)
    start := time.Now()

    refs, ok := r.URL.Query()["ref"]
    if !ok || len(refs) < 1 {
      http.Error(w, "Thunderzippy. Pass ?ref= to use.", 500)
      return
    }
    ref := refs[0]

    files, err := getFileListByZipReferenceId(ref)
    if err != nil {
      http.Error(w, err.Error(), 403)
      log.Printf("%s\t%s\t%s", r.Method, r.RequestURI, err.Error())
      return
    }

    w.Header().Add("Content-Disposition", "attachment; filename=\"documents-" + ref + ".zip\"")
    w.Header().Add("Content-Type", "application/zip")

    zipWriter := zip.NewWriter(w)

    for _, file := range files {
      addDownloadToZip(zipWriter, file.Url, file.Filepath)
    }

    err = zipWriter.Close()
    if err != nil {
      log.Fatal(err)
    }

    log.Printf("Thunderzipped:\t%d files (%s)", len(files), time.Since(start))
}

func ZipReferenceCreate(w http.ResponseWriter, r *http.Request) {
  var files []*ZipEntry

  body, err := ioutil.ReadAll(io.LimitReader(r.Body, 1048576))
  if err != nil {
    http.Error(w, "Data was too limited to allow", 403)
    return
  }

  if err := r.Body.Close(); err != nil {
    http.Error(w, "Wrong Data. Please try again", 403)
    return
  }

  if err := json.Unmarshal(body, &files); err != nil {
    w.Header().Set("Content-Type", "application/json; charset=UTF-8")
    w.WriteHeader(422)

    if err := json.NewEncoder(w).Encode(err); err != nil {
      http.Error(w, "Wrong Data. Please try again", 403)
      return
    }
  }

  ref_id := CreateZipReference(files)
  w.Header().Set("Content-Type", "application/json; charset=UTF-8")
  w.WriteHeader(http.StatusCreated)
  js, err := json.Marshal(map[string]interface{}{
    "ref": ref_id,
  })
  HandleError(err)
  w.Write(js)
}

func addDownloadToZip(zipWriter *zip.Writer, url string, name string) {
  // https://golang.org/pkg/net/http/
  resp, err := http.Get(url)
  if err != nil {
    log.Print(err)
    return
  }
  log.Printf("adding:\t%d %s", resp.StatusCode, url)
  if resp.StatusCode != 200 {
    return
  }
  defer resp.Body.Close()

  // https://golang.org/pkg/archive/zip/#FileHeader
  h := &zip.FileHeader{
    Name:   name,
    Method: zip.Deflate,
    Flags:  0x800,
  }
  h.SetModTime(time.Now())
  f, err := zipWriter.CreateHeader(h)
  if err != nil {
    log.Print(err)
    return
  }

  io.Copy(f, resp.Body)
}


thunderzippy.go
package main

import (
    "archive/zip"
    "encoding/json"
    "errors"
    redigo "github.com/garyburd/redigo/redis"
    "io"
    "log"
    "net/http"
    "os"
    "time"
)

var redisPool *redigo.Pool
var config = Configuration{}

type Configuration struct {
    RedisServerAndPort string
    Port               string
}

type ZipEntry struct {
    Filepath, Url string
}

func zipHandler(w http.ResponseWriter, r *http.Request) {
    log.Printf("%s\t\t%s", r.Method, r.RequestURI)
    start := time.Now()

    refs, ok := r.URL.Query()["ref"]
    if !ok || len(refs) < 1 {
        http.Error(w, "Thunderzippy. Pass ?ref= to use.", 500)
        return
    }
    ref := refs[0]

    files, err := getFileListFromRedis(ref)
    if err != nil {
        http.Error(w, err.Error(), 403)
        log.Printf("%s\t%s\t%s", r.Method, r.RequestURI, err.Error())
        return
    }

    w.Header().Add("Content-Disposition", "attachment; filename=\"download.zip\"")
    w.Header().Add("Content-Type", "application/zip")

    zipWriter := zip.NewWriter(w)

    for _, file := range files {
        addDownloadToZip(zipWriter, file.Url, file.Filepath)
    }

    err = zipWriter.Close()
    if err != nil {
        log.Fatal(err)
    }

    log.Printf("Thunderzipped:\t%d files (%s)", len(files), time.Since(start))
}

func addDownloadToZip(zipWriter *zip.Writer, url string, name string) {
    // https://golang.org/pkg/net/http/
    resp, err := http.Get(url)
    if err != nil {
        log.Print(err)
        return
    }
    log.Printf("adding:\t%d %s", resp.StatusCode, url)
    if resp.StatusCode != 200 {
        return
    }
    defer resp.Body.Close()

    // https://golang.org/pkg/archive/zip/#FileHeader
    h := &zip.FileHeader{
        Name:   name,
        Method: zip.Deflate,
        Flags:  0x800,
    }
    h.SetModTime(time.Now())
    f, err := zipWriter.CreateHeader(h)
    if err != nil {
        log.Print(err)
        return
    }

    io.Copy(f, resp.Body)
}

func getFileListFromRedis(ref string) (files []*ZipEntry, err error) {
    redis := redisPool.Get()
    defer redis.Close()

    // Get the value from Redis
    result, err := redis.Do("GET", "zip:"+ref)
    if err != nil || result == nil {
        err = errors.New("Access Denied (sorry your link has timed out)")
        return
    }

    // Convert to bytes
    var resultByte []byte
    var ok bool
    if resultByte, ok = result.([]byte); !ok {
        err = errors.New("Error converting data stream to bytes")
        return
    }

    // Decode JSON
    err = json.Unmarshal(resultByte, &files)
    if err != nil {
        err = errors.New("Error decoding json: " + string(resultByte))
    }

    return
}

func initRedis() {
    log.Printf("Opening Redis: %s", config.RedisServerAndPort)
    redisPool = &redigo.Pool{
        MaxIdle:     10,
        IdleTimeout: 1 * time.Second,
        Dial: func() (redigo.Conn, error) {
            return redigo.Dial("tcp", config.RedisServerAndPort)
        },
        TestOnBorrow: func(c redigo.Conn, t time.Time) (err error) {
            _, err = c.Do("PING")
            if err != nil {
                panic("Error connecting to redis")
            }
            return
        },
    }
}

func initConfig() {
    conf := "thunderzippy_conf.json"
    configFile, _ := os.Open(conf)
    log.Printf("Reading config %s", conf)
    decoder := json.NewDecoder(configFile)
    err := decoder.Decode(&config)
    if err != nil {
        panic("Error reading conf")
    }
}

func main() {
    initConfig()
    initRedis()
    log.Printf("Thunderzippy is go on port %s", config.Port)
    http.HandleFunc("/zip/", zipHandler)
    http.ListenAndServe(":"+config.Port, nil)
}
